### ISSUE TITLE
Add a way to do something on progress

### ISSUE THREAD

**Issue by fabi321** [DzenanJupic/rustube#6]
It would be nice to be able to do stuff on progress, like pytube provides the ability to add an `on_download_progress` callback. This would be helpful for displaying a progress bar.

**Comment by fabi321** [comment]
This could also be implemented as an async stream of information about the download.

**Comment by DzenanJupic** [comment]
@fabi321 sounds reasonable.

The best place for such a hook would probably be in `src/stream/mod.rs:261`
```rust
async fn write_stream_to_file(mut stream: impl tokio_stream::Stream<Item=reqwest::Result<bytes::Bytes>> + Unpin, file: &mut File) -> Result<()> {
        while let Some(chunk) = stream.next().await {
            file
                .write_all(&chunk?)
                .await?;
        }
        Ok(())
    }
```
This function handles the response stream and writes incoming byte-chunks to the target file.

Implementing it similar to `pytube`, by having a dedicated `struct` for download arguments is probably the best idea. This would allow adding arguments in the future without making a breaking change.

I won't have time to implement this within the next few days, so you're welcome to make PR. Otherwise, I will implement it within the next few weeks.

**Comment by DzenanJupic** [comment]
> This could also be implemented as an async stream of information about the download.

That also sounds like an interesting option, especially, since it allows the other side to stop listening when it's done with its task.

`tokio`s [channel](https://docs.rs/tokio/1.4.0/tokio/sync/mpsc/fn.channel.html) could be used for that. It returns a [Receiver](https://docs.rs/tokio/1.4.0/tokio/sync/mpsc/struct.Receiver.html) that can be turned into a `Stream` using [ReceiverStream](https://docs.rs/tokio-stream/0.1.5/tokio_stream/wrappers/struct.ReceiverStream.html).

The main question here would be if the download should stop when the `Receiver` is dropped - kind of like a kill switch - or if such errors should be ignored.

**Comment by fabi321** [comment]
I tried to hide this behind a feature since it needs tokio/sync, but the resulting code either has a lot of duplication or looks ugly. Would you mind if my solution does require a feature to be activated?

**Comment by fabi321** [comment]
Nvm, I just added yet more fields to the Stream struct.

**Comment by DzenanJupic** [comment]
@fabi321 thank you for the effort.

It's no problem if it's activated without a feature. The feature system is far from perfect/finished and I will eventually redo it from scratch.

**Comment by fabi321** [comment]
I now opened #9 as a first go to. It's using a feature now

**Comment by fabi321** [comment]
It's completely lacking an actual executor, but provides my first thoughts on the topic

**Comment by fabi321** [comment]
Fixed as #9 is merged
